providers/keycloak.go

```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
 	p.Group = group
 }
 
-func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	json, err := requests.New(p.ValidateURL.String()).
+func (p *KeycloakProvider) SetRoles(roles []string) {
+	p.Roles = roles
+}
+
+func extractRolesFromClaims(claims map[string]interface{}) ([]string, error) {
+	var roleList []string
+
+	if realmRoles, found := claims["realm_access"].(map[string]interface{}); found {
+		if roles, found := realmRoles["roles"]; found {
+			_, ok := roles.([]interface{})
+			if !ok {
+				return nil, errors.New("error parsing realm roles from claims")
+			}
+			for _, r := range roles.([]interface{}) {
+				roleList = append(roleList, fmt.Sprintf("%s", r))
+			}
+		}
+	}
+
+	if clientRoles, found := claims["resource_access"].(map[string]interface{}); found {
+		for name, list := range clientRoles {
+			scopes, ok := list.(map[string]interface{})
+			if !ok {
+				return nil, errors.New("error parsing client roles from claims")
+			}
+			if roles, found := scopes["roles"]; found {
+				for _, r := range roles.([]interface{}) {
+					roleList = append(roleList, fmt.Sprintf("%s:%s", name, r))
+				}
+			}
+		}
+	}
+
+	return roleList, nil
+}
+
+func (p *KeycloakProvider) getUserInfo(ctx context.Context, s *sessions.SessionState) (*simplejson.Json, error) {
+	userInfo, err := requests.New(p.ValidateURL.String()).
 		WithContext(ctx).
 		SetHeader("Authorization", "Bearer "+s.AccessToken).
 		Do().
 		UnmarshalJSON()
 	if err != nil {
 		logger.Errorf("failed making request %s", err)
-		return "", err
+		return nil, fmt.Errorf("error getting user info: %v", err)
 	}
+	return userInfo, nil
+}
 
-	if p.Group != "" {
-		var groups, err = json.Get("groups").Array()
+func getEmailAddress(userInfo *simplejson.Json) (string, error) {
+	var email, err = userInfo.Get("email").String()
+	if err != nil {
+		return "", fmt.Errorf("error getting email address: %v", err)
+	}
+	return email, nil
+}
+
+func getGroups(userInfo *simplejson.Json) ([]string, error) {
```

Comment:
Will we need to harmonize this groups scope creep with this related PR if it gets picked up again? https://github.com/oauth2-proxy/oauth2-proxy/pull/861

Reply:
I believe so.